python/social_bot/gazebo_agent.py

```diff
@@ -170,6 +190,25 @@ def get_observation(self, teacher, sentence_raw="hello"):
             obs = self.get_camera_observation()
         else:  # observation is pure low-dimentional states
             obs = teacher.get_task_specific_observation(self)
+            if self._goal_conditioned:
+                flat_obs = obs
+                obs = OrderedDict()
+                obs['observation'] = flat_obs
+                obs['achieved_goal'] = flat_obs[6:8]
+                obs['desired_goal'] = flat_obs[12:14]
+                if self._use_aux_achieved:
+                    obs['observation'] = flat_obs[14:]
+                    obs['aux_achieved'] = np.concatenate(
+                        (flat_obs[:6], flat_obs[8:12]), axis=0)
+                    if self._xy_only_aux:
+                        # 2: z-speed, 3, 4: angular velocity, 5: yaw-vel, 6, 7: x, y, 8: z, 9, 10, 11: pose
+                        obs['observation'] = np.concatenate(
+                            (flat_obs[2:5], flat_obs[8:11], flat_obs[14:]),
```

Comment:
Perhaps we should not use this style operation. I think a better way is to put them into GoalTask, since they are only for GoalTask so not supposed to be here, and GoalTask has the original data so we don't need to spilt numpy array:

```
if self._goal_conditioned:
        tasks = teacher.get_task_groups().get_tasks()
        assert_for_goal_conditioned_ob()
        tasks[0].get_goal_conditioned_ob()
```

And then add get_goal_conditioned_ob() in Goaltask, as well as the option self._use_aux_achieved and self._xy_only_aux to Task. We can also use it to get a sample observation to get feilds and shapes, so we don't have to use a fixed shape in "_construct_goal_condi_space()":


```
def construct_goal_condi_space:
        assert_for_goal_conditioned_ob
        get a sample observation from get_goal_conditioned_ob
        get feilds and shapes and set the space
```

```
def assert_for_goal_conditioned_ob:
        assert self.type == 'pioneer2dx_noplugin'
        assert not self._with_language
        assert not self._use_image_observation
        assert not self._image_with_internal_states
        tgs = teacher.get_task_groups()
        assert len(tgs) == 1
        tasks = tgs[0].get_tasks()
        assert len(tasks) == 1 and isinstance(tasks[0], GoalTask), str(tasks)

```

Reply:
Good points.  Done all.

Thanks for the comments, and PTAL.